aten/src/ATen/core/jit_type.h

```diff
@@ -1123,6 +1126,88 @@ struct CAFFE2_API TupleType : public NamedType {
   std::shared_ptr<FunctionSchema> schema_;
 };
 
+struct EnumType;
+using EnumTypePtr = std::shared_ptr<EnumType>;
+struct CAFFE2_API EnumType : public NamedType {
+  friend struct Type;
+  static const TypeKind Kind = TypeKind::EnumType;
+
+  static EnumTypePtr create(const c10::QualifiedName& qualified_name, TypePtr value) {
+    switch (value->kind()) {
+      case TypeKind::AnyType:
+      case TypeKind::IntType:
+      case TypeKind::FloatType:
+      case TypeKind::StringType:
+      case TypeKind::TensorType:
+        return EnumTypePtr(new EnumType(qualified_name, value));
+      default:
+        AT_ERROR(
+            "Cannot create Enum with value type '",
+            value->str(),
+            "', only int, float, Tensor and string keys are supported");
+    }
+  }
+
+  std::string str() const override {
+    return "Enum<" + annotation_str() + ">";
+  }
+
+  std::string repr_str() const override {
+    return str();
+  }
+
+  TypePtr getValueType() const {
+    return value_type;
+  }
+
+  bool operator==(const Type& rhs) const override {
+    if (auto enum_rhs = rhs.cast<EnumType>()) {
+      return name().value() == enum_rhs->name().value() &&
```

Comment:
I think we should probably be doing the same logic as https://github.com/pytorch/pytorch/blob/fc261692930cc161f4679a5c5a8ed06f093a2d94/aten/src/ATen/core/jit_type.h#L1858 here

Reply:
Done.